modules/peridynamics/src/kernels/FiniteStrainMechanicsNOSPD.C

```diff
@@ -0,0 +1,465 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FiniteStrainMechanicsNOSPD.h"
+#include "MeshBasePD.h"
+#include "RankTwoTensor.h"
+
+registerMooseObject("PeridynamicsApp", FiniteStrainMechanicsNOSPD);
+
+template <>
+InputParameters
+validParams<FiniteStrainMechanicsNOSPD>()
+{
+  InputParameters params = validParams<MechanicsBaseNOSPD>();
+  params.addClassDescription(
+      "Class for calculating residual and Jacobian for Self-stabilized Non-Ordinary "
+      "State-based PeriDynamic (SNOSPD) formulation under finite strain assumptions");
+
+  params.addRequiredParam<unsigned int>(
+      "component",
+      "An integer corresponding to the variable this kernel acts on (0 for x, 1 for y, 2 for z)");
+
+  return params;
+}
+
+FiniteStrainMechanicsNOSPD::FiniteStrainMechanicsNOSPD(const InputParameters & parameters)
+  : MechanicsBaseNOSPD(parameters),
+    _dgrad_old(getMaterialPropertyOld<RankTwoTensor>("deformation_gradient")),
+    _E_inc(getMaterialProperty<RankTwoTensor>("strain_increment")),
+    _R_inc(getMaterialProperty<RankTwoTensor>("rotation_increment")),
+    _component(getParam<unsigned int>("component"))
+{
+}
+
+void
+FiniteStrainMechanicsNOSPD::computeLocalResidual()
+{
+  // For finite strain formulation, the _stress tensor gotten from material class is the
+  // Cauchy stress (Sigma). the first Piola-Kirchhoff stress (P) is then obtained as
+  // P = J * Sigma * inv(F)^T.
+  // Nodal force states are based on the first Piola-Kirchhoff stress tensors (P).
+  // i.e., T = (J * Sigma * inv(F)^T) * inv(Shape) * xi * multi.
+  // Cauchy stress is calculated as Sigma_n+1 = Sigma_n + R * (C * dt * D) * R^T in the
+  // FiniteStrainNOSPD material class.
+
+  std::vector<RankTwoTensor> nodal_force(_nnodes);
+  for (unsigned int nd = 0; nd < _nnodes; ++nd)
+    nodal_force[nd] = (_dgrad[nd].det() * _stress[nd] * _dgrad[nd].inverse().transpose()) *
```

Comment:
@hchen139 Referring to your other comment, where does the 'multi' here for the finite strain mechanics get calculated ?